telegraf.go

```diff
@@ -60,318 +63,213 @@ type TelegrafConfigFilter struct {
-		tc.ID = *tcd.ID
-	}
-	if orgID := tcd.OrgID; orgID != nil && orgID.Valid() {
-		tc.OrgID = *orgID
-	} else if tcd.OrganizationID != nil {
-		tc.OrgID = *tcd.OrganizationID
-	}
-	return decodePluginRaw(tcd, tc)
-}
-
-func decodePluginRaw(tcd *telegrafConfigDecode, tc *TelegrafConfig) (err error) {
+// return bucket, config, error
+func decodePluginRaw(tcd *telegrafConfigDecode) ([]string, string, error) {
 	op := "unmarshal telegraf config raw plugin"
-	for k, pr := range tcd.Plugins {
+	ps := ""
+	bucket := []string{}
+
+	for _, pr := range *tcd.Plugins {
 		var tpFn func() plugins.Config
-		var config plugins.Config
 		var ok bool
+
 		switch pr.Type {
-		case plugins.Input:
+		case "input":
 			tpFn, ok = availableInputPlugins[pr.Name]
-		case plugins.Output:
+		case "output":
 			tpFn, ok = availableOutputPlugins[pr.Name]
 		default:
-			return &Error{
+			return nil, "", &Error{
 				Code: EInvalid,
-				Msg:  fmt.Sprintf(ErrUnsupportTelegrafPluginType, pr.Type),
 				Op:   op,
+				Msg:  fmt.Sprintf(ErrUnsupportTelegrafPluginType, pr.Type),
 			}
 		}
-		if ok {
-			config = tpFn()
-			// if pr.Config if empty, make it a blank obj,
-			// so it will still go to the unmarshalling process to validate.
-			if len(string(pr.Config)) == 0 {
-				pr.Config = []byte("{}")
-			}
-			if err = json.Unmarshal(pr.Config, config); err != nil {
-				return &Error{
-					Code: EInvalid,
-					Err:  err,
-					Op:   op,
-				}
+
+		if !ok {
+			return nil, "", &Error{
+				Code: EInvalid,
+				Op:   op,
+				Msg:  fmt.Sprintf(ErrUnsupportTelegrafPluginName, pr.Name, pr.Type),
 			}
-			tc.Plugins[k] = TelegrafPlugin{
-				Comment: pr.Comment,
-				Config:  config,
+		}
+
+		config := tpFn()
+		// if pr.Config if empty, make it a blank obj,
+		// so it will still go to the unmarshalling process to validate.
+		if len(string(pr.Config)) == 0 {
+			pr.Config = []byte("{}")
+		}
+
+		if err := json.Unmarshal(pr.Config, config); err != nil {
+			return nil, "", &Error{
+				Code: EInvalid,
+				Err:  err,
+				Op:   op,
 			}
-			continue
 		}
-		return &Error{
-			Code: EInvalid,
-			Op:   op,
-			Msg:  fmt.Sprintf(ErrUnsupportTelegrafPluginName, pr.Name, pr.Type),
+
+		if pr.Name == "influxdb_v2" {
+			if b := config.(*outputs.InfluxDBV2).Bucket; b != "" {
+				bucket = []string{b}
+			}
 		}
 
+		ps += config.TOML()
 	}
-	return nil
+
+	return bucket, ps, nil
+}
+
+// telegrafConfigDecode is the helper struct for json decoding. legacy.
+type telegrafConfigDecode struct {
+	ID             *ID                     `json:"id,omitempty"`
+	OrganizationID *ID                     `json:"organizationID,omitempty"`
+	OrgID          *ID                     `json:"orgID,omitempty"`
+	Name           string                  `json:"name,omitempty"`
+	Description    string                  `json:"description,omitempty"`
+	Config         string                  `json:"config,omitempty"`
+	Plugins        *[]telegrafPluginDecode `json:"plugins,omitempty"`
```

Comment:
Why is this a pointer to a slice?

Reply:
Likely so `Plugins` will be `nil` if the json payload is patterned after the new structure. I'm open to suggestions

Reply:
the zero value for a slice, is `nil`, and will be omitted by the json marshaler in the case it is a zero value of `nil`, the pointer shouldn't be necessary  🤔